### ISSUE TITLE
Повысить читабельность кода 

### ISSUE THREAD

**Issue by stanislav-az** [unam3/trial-bot-all#7]
Сейчас некоторые функции имеют очень большой уровень вложенности, из-за чего сложно понимать что эти функции из себя представляют. Поскольку код в первую очередь пишется для людей нужно представлять как бы другой человек смог разобраться в написанной функции, и уменьшать сложность кода.

Для уменьшения сложности есть следующие подходы:
1. **Разделяй** большую функцию на функции поменьше. Итоговая проблема которую решала функция разделится при этом на более мелкие проблемы и читающему код станет легче поскольку в голове в каждый момент чтения можно будет держать меньше информации. Это вообще классический принцип, он помогает не только понимать программы, но и в принципе их писать, я думаю ты о нем слышал, есть даже на вики [статья](https://ru.wikipedia.org/wiki/%D0%A0%D0%B0%D0%B7%D0%B4%D0%B5%D0%BB%D1%8F%D0%B9_%D0%B8_%D0%B2%D0%BB%D0%B0%D1%81%D1%82%D0%B2%D1%83%D0%B9_(%D0%B8%D0%BD%D1%84%D0%BE%D1%80%D0%BC%D0%B0%D1%82%D0%B8%D0%BA%D0%B0)). 
2. Нужно постараться сделать так чтобы у каждой функции была одна зона ответственности, чтобы у нее была **одна четкая цель**. Тоже стандартный термин для этого применяется - [связность](https://en.wikipedia.org/wiki/Cohesion_(computer_science)).
3. **Используй стандартные паттерны проектирования** вместо того чтобы заново их реализовывать самому. Например, для обработки списков есть функции `filter`, `map`, `foldr` (есть также для списков с эффектами аналоги). Не стоит использовать рекурсию, а вместо этого лучше выразить то же через стандартные функции. Поскольку каждый их встречал и использовал, читающему будет легче понять смысл функции. Заметил подобный код в `Telegram.HandleUpdates.handleUpdates`.
4. **Уменьшить вложенность кода**. Каждый дополнительный уровень приводит к тому, что при чтении необходимо помнить все вложенные конструкции вверх по "стеку".

**Comment by unam3** [comment]
> Сейчас некоторые функции имеют очень большой уровень вложенности, из-за чего сложно понимать что эти функции из себя представляют.
…
Разделяй большую функцию на функции поменьше.


Огласите пожалуйста названия таких функций.

> Не стоит использовать рекурсию, а вместо этого лучше выразить то же через стандартные функции.

Было здорово прикладывать название функций/ссылку на участки кода где такое используется.


> Заметил подобный код в Telegram.HandleUpdates.handleUpdates.

Точно в этом проекте? `git grep -in HandleUpdates` ничего не выдает.

**Comment by stanislav-az** [comment]
> Точно в этом проекте? git grep -in HandleUpdates ничего не выдает.

 Да, действительно, случайно забыл поправить название функций. Имел ввиду `cycleLPProcessing` и вариации `cycleEcho`. В модулях `src/Vk.hs` и `src/Tg.hs`.
 
 > Огласите пожалуйста названия таких функций.

Это к примеру `processUpdates` в модулях `src/Vk.hs` и `src/Tg.hs`.

**Comment by stanislav-az** [comment]
P.S. Извини за долгий ответ, я был в отпуске месяц где-то, поэтому не ответил

**Comment by unam3** [comment]
>> Точно в этом проекте? git grep -in HandleUpdates ничего не выдает.


> Да, действительно, случайно забыл поправить название функций. Имел ввиду cycleLPProcessing и вариации cycleEcho. В модулях src/Vk.hs и src/Tg.hs.

На пальцах изобразите как это может выглядеть. Не могу представить замену рекурсии в функции с ветвлениями:
https://github.com/unam3/trial-bot-all/blob/e08b56116c4b1ee3136e38d53d3122a2d2404e53/src/Vk.hs#L102-L121
